File: samples/gateway_demo.php
<?php

declare(strict_types=1);

/**
 * Sample PHP handlers — mirror of samples/gateway_demo.py.
 *
 * Used by mod_php's CLI via `--bootstrap`; also imported by the
 * tests/test_gateway_e2e_php.py end-to-end test that spawns
 * mod_php as a subprocess and verifies the round-trip.
 *
 * Run end-to-end:
 *
 *     # Terminal 1
 *     python -m server --gateway-socket /tmp/agtpd.sock \
 *         --agents-dir server/agents --endpoints-dir endpoints
 *
 *     # Terminal 2
 *     php mod_php/bin/run.php \
 *         --gateway-socket /tmp/agtpd.sock \
 *         --bootstrap samples/gateway_demo.php
 *
 * The composer.json at agtp-php/ uses path-style autoloading; the
 * host project autoloader (or a separate `composer install` inside
 * mod_php/) provides PSR-4 resolution for the Agtp\ namespace.
 */

namespace Samples\GatewayDemo;

use Agtp\AgtpEndpoint;
use Agtp\EndpointContext;
use Agtp\EndpointError;
use Agtp\EndpointResponse;
use Agtp\HandlerRegistry;

final class GatewayDemoHandlers
{
    #[AgtpEndpoint(method: 'QUERY', path: '/echo')]
    public function echo(EndpointContext $ctx): EndpointResponse
    {
        return new EndpointResponse(body: [
            'echo' => (string) ($ctx->input['value'] ?? ''),
        ]);
    }

    #[AgtpEndpoint(
        method: 'BOOK',
        path: '/room',
        errors: ['room_unavailable'],
    )]
    public function bookRoom(EndpointContext $ctx): EndpointResponse|EndpointError
    {
        $roomType = (string) ($ctx->input['room_type'] ?? 'double');
        if ($roomType === 'presidential_suite') {
            return new EndpointError(
                code: 'room_unavailable',
                message: 'The presidential suite is not available.',
                details: ['room_type' => $roomType],
            );
        }
        return new EndpointResponse(body: [
            'reservation_id' => 'res-' . (string) ($ctx->input['guest'] ?? 'anon') . '-' . $roomType,
            'agent' => $ctx->agentId,
        ]);
    }
}

HandlerRegistry::default()->registerClass(GatewayDemoHandlers::class);

File: samples/gateway_demo.py
"""
Sample handlers demonstrating the M3 step (b) ``@endpoint`` decorator.

Run alongside ``agtpd --gateway-socket`` to see end-to-end gateway
dispatch in action::

    # terminal 1
    python -m server --gateway-socket /tmp/agtpd.sock \\
        --agents-dir server/agents --endpoints-dir endpoints

    # terminal 2
    python -m mod_python \\
        --gateway-socket /tmp/agtpd.sock \\
        --load-module samples.gateway_demo

When agtpd's TOML endpoints reference these functions by dotted path
(``samples.gateway_demo.echo``, ``samples.gateway_demo.book_room``),
the daemon sends the references in the ``register`` frame and this
module resolves them against ``agtp.registry``.

These handlers are intentionally tiny — the point is to exercise the
gateway round-trip, not to demonstrate AGTP modeling.
"""

from __future__ import annotations

from agtp import EndpointContext, EndpointError, EndpointResponse, endpoint


@endpoint(method="QUERY", path="/echo")
def echo(ctx: EndpointContext) -> EndpointResponse:
    """Return the input ``value`` unchanged. The simplest possible handler."""
    return EndpointResponse(body={"echo": ctx.input.get("value", "")})


@endpoint(
    method="BOOK",
    path="/room",
    errors=["room_unavailable"],
)
def book_room(ctx: EndpointContext):
    """Toy room-booking handler. Refuses the presidential suite."""
    room_type = ctx.input.get("room_type", "double")
    if room_type == "presidential_suite":
        return EndpointError(
            code="room_unavailable",
            message="The presidential suite is not available.",
            details={"room_type": room_type},
        )
    return EndpointResponse(body={
        "reservation_id": f"res-{ctx.input.get('guest', 'anon')}-{room_type}",
        "agent": ctx.agent_id,
    })
